StyleSelect 구현

Hansolkkim opened the pull request:
StyleSelect 도메인 구현을 일단 완료했습니다.
현재 이미지(UIImage) 위에 특정 Font를 가지는 Text를 올리는 정도로만 Style을 사용하고 있습니다.
추후 기능을 더 추가하여 다른 모양이나 배치 등을 달리하는 것을 Style로 변경할 수 있도록 할 예정입니다.

Hansolkkim committed:
feat: [#68] StyleSelectView에서 사용할 Style 타입 추가

- 현재 Style == Font 로만 우선적으로 고려하여 Style 변경 시 이미지에 올라갈 Text의 UIFont를 변경하도록, Style 타입이 Font Name을 가지고 있도록 설정
- Photo 타입이 Style 타입의 정보를 가지고 있도록, Photo 타입에 style 프로퍼티 추가

Hansolkkim committed:
feat: [#68] StyleSelect 도메인 구현

- PhotoEditViewModel에서 StyleSelectViewModel에게 PublishRelay<BehaviorRelay<Photo>>의 참조 타입 프로퍼티를 넘겨주도록 설정
- UIImage를 extension하여, UIImage 위에 정해진 Color, Style의 문구가 표시되도록 하는 메소드 추가

Hansolkkim committed:
fix: [#68] ColorSelectView의 BottomSheet 관련 로직 수정

- 변경된 BottomSheet에 맞게 ColorSelectView의 BottomSheet 관련 로직 수정

Hansolkkim committed:
feat: [#68] PhotoEditView와 AlbumCell의 Image에 Text가 추가되어 보이도록 설정

Hansolkkim committed:
feat: [#68] StyleSelectCollectionView와 AlbumCollectionView의 Cell 터치 관련 로직 추가

- StyleSelectCollectionView와 AlbumCollectionView의 Cell 터치시, 터치된 Cell이 CenteredHorizontally하게 화면이 변경되도록 설정

Hansolkkim commented on rabit/rabit/Album/AlbumCoordinator.swift:
```diff
@@ -10,29 +10,32 @@ protocol AlbumNavigation {
 protocol PhotoEditNavigation {
     var showColorSelectView: PublishRelay<BehaviorRelay<String>> { get }
-    var showStylePickerView: PublishRelay<Void> { get }
+    var showStyleSelectView: PublishRelay<BehaviorRelay<Album.Item>> { get }
```
StyleSelectView의 StyleSelectCollectionView에서는 Cell에 적용된 Text Style을 보여줌으로써 선택할 수 있게하기 위해, BehaviorRelay<Album.Item> 타입을 넘겨주도록 했습니다.

(ColorSelectCollectionView에서는 따로 예시를 보여줄 필요가 없다고 생각해 BehaviorRelay\<Color\> 타입을 넘겨주도록 했었습니다.)

Hansolkkim commented on rabit/rabit/Album/AlbumCoordinator.swift:
```diff
@@ -10,29 +10,32 @@ protocol AlbumNavigation {
 
 protocol ColorSelectNavigation {
     var closeColorSelectView: PublishRelay<Void> { get }
-    var saveSelectedColor: PublishRelay<Void> { get }
```
얘를 기존에는 ColorSelectViewModel이 PhotoEditViewModel에게 "사진이 저장됐음" 을 알려주어서 PhotoEditView의 ImageView를 다시 그리기 위해 사용했었는데, 현재 ColorStream을 이용하여 ColorSelectView의 저장 버튼이 눌리면 자동으로 ImageView가 다시 그려지기 때문에 필요가 없어져 제거했습니다!
(사실 왜 아직까지 남아있는지 잘 모르겠어요..ㅋㅋ)

junu0516 replied:
이게 기존 뷰모델에서 가지고 있던 스트림을 전달받아서 다른 뷰모델이 구독하기 때문에 결과적으로 두 뷰모델에 값이 다 방출되는거죠!? 
ㅋㅋ 뷰모델 계속 수정하다보면 나중에 오히려 필요없어지는 부분이 생기는거 같아요

Hansolkkim replied:
오 조리있게 잘 정리해주셨네요ㅋㅋㅋㅋ제가 드리려던 말씀이 그거였습니다ㅎㅎ

Hansolkkim commented on rabit/rabit/Album/Models/Style.swift:
```diff
@@ -0,0 +1,43 @@
+    case Verdana
+    case none
+
+    init(_ StyleString: String) {
+        switch StyleString.lowercased() {
+        case "helvetica": self = .Helvetica
+        case "americantypewriter": self = .AmericanTypewriter
+        case "applesdgothic": self = .AppleSDGothic
+        case "noteworthy": self = .Noteworthy
+        case "gillsans": self = .GillSans
+        case "verdana": self = .Verdana
+        default: self = .none
+        }
+    }
```
PhotoEntity는 Style을 String타입으로 가지고 있기 때문에, PhotoEntity -> Photo 변환시에 Style 타입을 생성해야하는 경우가 생겼습니다.
이를 위해 초기화 생성자를 두었습니다.

junu0516 replied:
PhotoEntity 생성 시에는 StyleString 형태로 속성을 가지고 있고, Photo 생성 시에는 StyleString으로 Style 열거형을 생성해서 속성을 가지는거죠!!? 이해했습니다 ㅋㅋ

추가로 저는 이전에 열거형 생성할 때 기본으로 제공되는 rawValue를 파라미터로 받는 실패가능한 생성자를 사용했는데, 이렇게 그냥 제가 생성자 정의해서 쓰는게 옵셔널바인딩이 필요없어서 더 편리하겠네요ㅎㅎ

Hansolkkim replied:
오 맞아요ㅋㅋ따로 뭐 none Case 같은걸 만들어서 초깃값처럼 사용하든지, 오류로 처리하든지 하면 편하더라고요!

Hansolkkim commented on rabit/rabit/Album/PhotoEdit/PhotoEditViewModel.swift:
```diff
@@ -79,14 +82,9 @@ private extension PhotoEditViewModel {
-                    date: $1.date,
-                    color: $0
-                )
+                var photo = $1
+                photo.color = $0
+                return photo
```
이 부분은 Photo 모델 타입의 color를 var로 선언된 변수이기 때문에, 새로 Photo 모델을 생성해주는 것이 아닌 color 값만 바꾸도록 하기 위해 위처럼 수정했습니다.

Hansolkkim replied:
삭제된 코드들은 보이지가 않아서 이상해 보이는 코멘트네요ㅋㅋㅋ [여기](https://github.com/GunZaStation/Rabit/pull/69/files/2558c3f1b073e97d851ea2472250136b4cd0e46c#diff-5ba6bad297ffff82c59c141a2c7450f2ef361b9a79c6a6b21d6852f007792636) 들어가서 보시면 됩니다!

junu0516 replied:
ㅎㅎ 더 간단해져서 보기좋네요!

Hansolkkim commented on rabit/rabit/Album/StyleSelect/StyleSelectViewController.swift:
```diff
@@ -0,0 +1,201 @@
+            .bind(to: viewModel.appliedPhotoWithSelectedStyle)
+            .disposed(by: disposeBag)
+
+        styleSelectCollectionView.rx.itemSelected
+            .map { ($0, UICollectionView.ScrollPosition.centeredHorizontally, true) }
+            .bind(onNext: styleSelectCollectionView.scrollToItem(at:at:animated:))
+            .disposed(by: disposeBag)
```
타임스탬프 앱을 참고해보니, 가운데 있지 않는 Style Cell을 터치하면 해당 Cell이 가운데로 오도록 하는 기능이 있길래 참고해서 위 코드로 구현했습니다.ㅎㅎ

Hansolkkim committed:
chore: [#68] Photo 모델 정렬 후 반환하도록 코드 추가

Hansolkkim committed:
refactor: [#68] Observable.share()로 불필요한 Observable 중복 생성 방지

junu0516 commented on rabit/rabit/Album/PhotoEdit/PhotoEditViewController.swift:
```diff
@@ -102,22 +102,29 @@ private extension PhotoEdtiViewController {
-                viewController.photoImageView.image = image
-                viewController.resizePhotoImageView(with: image)
-            }
+        let updatedSelectedPhotoData = viewModel.selectedPhotoData
+                                            .map(\.imageData)
+                                            .compactMap(UIImage.init(data:))
+                                            .withLatestFrom(viewModel.selectedPhotoData) {
+                                                $0.overlayText(of: $1)
+                                            }
+                                            .compactMap { $0 }
+                                            .share()
+
+        updatedSelectedPhotoData
+            .bind(to: photoImageView.rx.image)
+            .disposed(by: disposeBag)
+
+        updatedSelectedPhotoData
+            .bind(onNext: resizePhotoImageView(with:))
```
`share()` 사용하는 이유가 하나의 시퀀스를 다른 옵저버들이 공유(?)하기 위해서 사용한다고 학습했는데,
이게 딱 여기 해당하는 예시..겠죠?? (아직 학습중이라 ㅋㅋ 이게 맞나 궁금해서 코멘트했습니다)

Hansolkkim replied:
맞습니다 
그 목적으로 사용하려했는데, 
앞서 구독하고 있는 image 변경 구독자가 너무 빨리 disposed 돼버리는 바람에 
뒤이어 구독하고 있는 resize용 구독자가 updatedSelectedPhotoData의 이벤트 방출을 받지 못하는 문제가 발생했어요ㅠㅠ 
조금 더 살펴보고 수정 후 머지하겠습니다!

Hansolkkim committed:
fix: [#68] Build Phase에 중복으로 존재하는 파일 제거

junu0516 commented on rabit/rabit/Album/Repository/AlbumRepository.swift:
```diff
@@ -46,8 +46,12 @@ private extension AlbumRepository {
+                filter: "categoryTitle == '\(categoryTitle)'"
+            )
                 .map(Photo.init)
+                .sorted { $0.date < $1.date }
```
Realm에서 데이터 리스트로 불러올 때 정렬 안되서 정렬하신거죠??
저도 GoalList 정렬하는거 처리 안했는데 같이 해야겠네요 ㅎㅎ

Hansolkkim replied:
네네ㅋㅋ기준 없이 저장된(제가 Realm에 입력했던) 순서대로 AlbumColllectionView가 보이길래, `Photo.date`를 기준으로 최근 찍은 사진이 제일 앞에 나오도록 해봤어요ㅎㅎ

junu0516 commented on rabit/rabit/Album/StyleSelect/StyleSelectViewController.swift:
```diff
@@ -0,0 +1,201 @@
+        return collectionView
+    }()
+
+    private let saveButton: UIButton = {
+        let button = UIButton(type: .system)
+        button.setTitle("저장", for: .normal)
+        button.titleLabel?.textAlignment = .center
+        button.titleLabel?.font = .systemFont(ofSize: 17, weight: .semibold)
+        button.setTitleColor(UIColor.white, for: .normal)
+        button.roundCorners()
+        button.setBackgroundColor(UIColor(named: "third"), for: .normal)
+        button.setBackgroundColor(.systemGray3, for: .disabled)
+        return button
+    }()
```
규모가 조금 커지다보니 드는 생각인데,
BottomSheet 형태를 우리가 공유하듯이 저장,취소,닫기 버튼 등 자주 사용될만한 컴포넌트를 또 하나 Common 하위에 지정해서 가지고 있어도 괜찮을 것 같아요!
특히 시간 및 날짜 선택화면 둘다 BottomSheet을 사용하는데 저장 버튼이 제각기 달리 존재해서 크기나 다른 형태도 달리 보이는데 애초에 이걸 전역 범위에서 공유되는 팩토리가 리턴하는 버튼을 사용하는 식으로 한 번 같이 진행해보는건 어떠세요!!?

간단히 말해서 우리가 공통으로 사용할만한 뷰를 팩토리에서 생성해서 주입받는 방식을 사용하는건 어떨까 싶었습니다!

Hansolkkim replied:
음 좋은 생각인것 같아요!! 안그래도 저희 버튼 대부분 색깔 통일해서 사용하고 있고, 그런 부분이 있어서 버튼팩토리 찬성입니다ㅎㅎ더 간단해질거 같아요

junu0516 commented on rabit/rabit/Goal/PeriodSelect/PeriodSelectViewModel.swift:
```diff
@@ -67,8 +67,12 @@ private extension PeriodSelectViewModel {
 
-        let startDate = usecase.startDate.compactMap { $0 }
-        let endDate = usecase.endDate.compactMap { $0 }
+        let startDate = usecase.startDate
+            .compactMap { $0 }
+            .share()
+        let endDate = usecase.endDate
+            .compactMap { $0 }
+            .share()
```
여기서의 `share()` 사용의 의미는 어떤건지 궁금합니다!!

Hansolkkim replied:
아 이 부분 앞서 코멘트주셨던 `updatedSelectedPhotoData`랑 다르게 동작하지 싶어서 추가했는데, 커밋에 포함됐네요..ㅋㅋ이 부분도 제거하고 머지하겠습니다ㅎㅎ

junu0516 reviewed (approved):
수고 많으셨어요 ㅎㅎ 궁금한점 위주로 코멘트 달았습니다.

추가로 코멘트에도 있는데 팩토리패턴으로 공통으로 사용할 뷰를 주입받는 방식을 같이 적용해보면 어떨까 싶어서 제안했습니다!! 다음에 오프라인으로 만날 때 이거 한 번 적용할만한지 회의하는거 어떠십니까 ㅎㅎ

Hansolkkim committed:
chore: [#68] 불필요하게 커밋에 추가된 부분 제거

Hansolkkim committed:
fix: [#68] share로 인한 문제 해결

- 기존에 share 오퍼레이터를 사용했을 때, 원인은 알 수 없지만 두 구독자들의 subscribe 타이밍에 차이가 있어 한 구독자는 이벤트를 받지 못 하는 이슈 발생
- share 오퍼레이터를 활용하려고 했던 목적을 생각했을 때, replay 매개변수를 활용함이 가장 적절하다고 판단하여 해당 방법으로 문제를 해결
